sockeye/data_io.py

```diff
@@ -461,15 +497,33 @@ def _assign_to_buckets(self, source_sentences, target_sentences):
             self.data_source[buck_idx].append(buff_source)
             self.data_target[buck_idx].append(buff_target)
             self.data_label[buck_idx].append(buff_label)
+            self.data_label_averge_len[buck_idx] += len(target)
+
+        # Average number of non-padding elements in target sequence per bucket
+        for buck_idx, buck in enumerate(self.buckets):
+            # Case of empty bucket -> use default padded length
+            if self.data_label_averge_len[buck_idx] == 0:
+                self.data_label_averge_len[buck_idx] = buck[1]
+            else:
+                self.data_label_averge_len[buck_idx] /= len(self.data_label[buck_idx])
+
+        # We now have sufficient information to populate bucket batch sizes
+        self._populate_bucket_batch_sizes()
 
         logger.info("Source words: %d", tokens_source)
         logger.info("Target words: %d", tokens_target)
         logger.info("Vocab coverage source: %.0f%%", (1 - num_of_unks_source / tokens_source) * 100)
         logger.info("Vocab coverage target: %.0f%%", (1 - num_of_unks_target / tokens_target) * 100)
-        logger.info('Total: {0} samples in {1} buckets'.format(len(self.data_source), len(self.buckets)))
+        logger.info("Total: %d samples in %d buckets", sum(len(b) for b in self.data_source), len(self.buckets))
         nsamples = 0
-        for bkt, buck in zip(self.buckets, self.data_source):
-            logger.info("bucket of {0} : {1} samples".format(bkt, len(buck)))
+        for bkt, buck, (batch_size_seq, _), average_seq_len in zip(
+                self.buckets, self.data_source, self.bucket_batch_sizes, self.data_label_averge_len):
+            logger.info("Bucket of %s : %d samples in %d batches of %d, approx %0.1f words/batch",
+                        bkt,
+                        len(buck),
+                        math.ceil(len(buck) / batch_size_seq),
+                        batch_size_seq,
+                        batch_size_seq * average_seq_len)
```

Comment:
We now report both sentences and average words per batch.  This really clarifies how unbalanced sentence-based batches can be.